examples/hyperpars_nb/script.jl

```diff
@@ -0,0 +1,344 @@
+using LinearAlgebra
+using Optim
+using ParameterHandling
+using Random
+using Statistics
+using StatsBase
+using Zygote
+
+# ## Global model settings
+kernel(θ) = θ.variance * SEKernel() ∘ ScaleTransform(θ.invlengthscale)
+likelihood(θ) = NegativeBinomialLikelihood(NBParamFailure(θ.r))
+
+# We subtract the logarithm of the likelihood parameter from the
+# mean. This accomplishes the same as if the likelihood were para-
+# metrized to yield E[y|f] = exp(f), and makes E[y] independent of
+# r.
+build_f(θ) = GP(θ.mean - log(θ.r), kernel(θ))
+build_latent_gp(θ) = LatentGP(build_f(θ), likelihood(θ), 1e-6)
+
+# ## ELBO
+# We define abstractions for holding the data for optimization
+# and methods to initialize, evaluate, and optimize the loss
+
+abstract type AbstractLoss end
+
+function optimize(l::AbstractLoss)
+    θ0 = init(l)
+    θ0_flat, unflatten = ParameterHandling.flatten(θ0)
+    unpack = ParameterHandling.value ∘ unflatten
+
+    opt = Optim.optimize(
+        l ∘ unpack,
+        θ -> only(Zygote.gradient(l ∘ unpack, θ)),
+        θ0_flat,
+        Optim.BFGS(;
+            alphaguess=Optim.LineSearches.InitialStatic(; scaled=true),
+            linesearch=Optim.LineSearches.BackTracking(),
+        ),
+        inplace=false,
+    )
+    display(opt)
+    return unpack(opt.minimizer)
+end
+
+function init_(l::AbstractLoss)
+    θ0 = init(l)
+    θ0_flat, unflatten = ParameterHandling.flatten(θ0)
+    unpack = ParameterHandling.value ∘ unflatten
+    return unpack(θ0_flat)
+end
+
+function (l::AbstractLoss)()
+    θ = init_(l)
+    return l(θ)
+end
+
+# This function will construct the posterior from the data and parameters.
+# Implementations below.
+function build_posterior(l::AbstractLoss, θ) end
+
+# This function estimates initial values for the mean and variance of the
+# latent function f, based on the marginal mean and variance of y, assuming
+# that the data came from a latent GP with Poisson likelihood.
+# This will be used below in the initializations of the optimizer.
+function initial_μ_σ²_Poisson(mean_y, var_y)
+    var_y < mean_y && return zero(mean_y), one(var_y)
+    σ² = log((var_y / mean_y - 1) / mean_y + 1)
+    μ = log(mean_y) - σ²/2
+    return μ, σ²
+end
+
+function correlation_length(y)
+    logvar = log(autocov(y, [0])[1])
+    i0 = length(y)
+    for i in 0:length(y)
+        ac = autocov(y, [i])[1]
+        if log(ac) - logvar < -0.5
+            i0 = i
+            break
+        end
+    end
+    return i0
+end
+
+# ## Diagonal SVGP model
+# We define a standard diagonal sparse variational GP for comparison
+
+struct DSVGPLoss{Tx, Ty} <: AbstractLoss
+    x::Tx
+    y::Ty
+end
+
+function init(l::DSVGPLoss)
+    μ, σ² = initial_μ_σ²_Poisson(mean(l.y), var(l.y))
+    xmin = minimum(l.x)
+    xmax = maximum(l.x)
+    Δt = (xmax - xmin) / length(l.x)
+    invl0 = inv(correlation_length(l.y) * Δt)
+    r0 = 100.
+    M = min(30, length(l.y))
+    z0 = rand(M)
+    z0 .*= xmax - xmin
+    z0 .+= xmin
```

Comment:
The reason why I'm interested in the augmented GP framework is to make training faster. The sparse variational GP is already much slower for moderate data sizes. If we make the variational GP use all training points, it would be hopelessly slow, wouldn't it?